src/gui/PluginBrowser.cpp

```diff
@@ -60,23 +61,89 @@ PluginBrowser::PluginBrowser( QWidget * _parent ) :
 								m_view );
 	hint->setWordWrap( true );
 
-	QScrollArea* scrollarea = new QScrollArea( m_view );
-	PluginDescList* descList = new PluginDescList( m_view );
-	scrollarea->setWidget(descList);
-	scrollarea->setWidgetResizable(true);
+	QLineEdit * searchBar = new QLineEdit( m_view );
+	searchBar->setPlaceholderText( "Search" );
+	searchBar->setMaxLength( 64 );
+	searchBar->setClearButtonEnabled( true );
 
-	view_layout->addWidget(hint);
-	view_layout->addWidget(scrollarea);
+	m_descTree = new QTreeWidget( m_view );
+	m_descTree->setColumnCount( 1 );
+	m_descTree->header()->setVisible( false );
+	m_descTree->setIndentation( 10 );
+	m_descTree->setSelectionMode( QAbstractItemView::NoSelection );
+
+	connect( searchBar, SIGNAL( textEdited( const QString & ) ),
+			this, SLOT( onFilterChanged( const QString & ) ) );
+
+	view_layout->addWidget( hint );
+	view_layout->addWidget( searchBar );
+	view_layout->addWidget( m_descTree );
+
+	// Add LMMS root to the tree
+	m_lmmsRoot = new QTreeWidgetItem();
+	m_lmmsRoot->setText( 0, "LMMS" );
+	m_descTree->insertTopLevelItem( 0, m_lmmsRoot );
+	m_lmmsRoot->setExpanded( true );
+
+	// Add LV2 root to the tree
+	m_lv2Root = new QTreeWidgetItem();
+	m_lv2Root->setText( 0, "LV2" );
+	m_descTree->insertTopLevelItem( 1, m_lv2Root );
+
+	// Add plugins to the tree roots
+	addPlugins();
+
+	// Resize
+	m_descTree->header()->setSectionResizeMode( QHeaderView::ResizeToContents );
+
+	// Hide empty roots
+	updateRootVisibilities();
+}
+
+
+void PluginBrowser::updateRootVisibility( int rootIndex )
+{
+	QTreeWidgetItem * root = m_descTree->topLevelItem( rootIndex );
+	if( root->childCount() )
+		root->setHidden( false );
```

Comment:
There are multiple uses of `if` without `{}`. Unfortunatelly, the LMMS coding conventions force us to use curly brackets.